### ISSUE TITLE
BlockingIOError: [Errno 35] Resource temporarily unavailable

### ISSUE THREAD

**Issue by StabbarN** [zopefoundation/zc.lockfile#23]
Any idea what's causing this?
```
.venv/lib/python3.7/site-packages/zc/lockfile/__init__.py:59: in _lock_file
    fcntl.flock(file.fileno(), _flags)
E   BlockingIOError: [Errno 35] Resource temporarily unavailable
```

More info at https://github.com/pytest-dev/pytest-services/issues/38

**Comment by mgedmin** [comment]
Errno 35 is EDEADLK.  The manual page for [fcntl(2)](http://manpages.ubuntu.com/manpages/focal/en/man2/fcntl.2.html) says

>  When  placing  locks with **F_SETLKW**, the kernel detects deadlocks, whereby two or more processes have their lock requests mutually blocked by locks held by the other processes.  For example, suppose process A holds a write lock on  byte  100  of  a  file,  and process B holds a write lock on byte 200.  If each process then attempts to lock the byte already locked by the other process using **F_SETLKW**, then, without deadlock detection, both processes would remain blocked indefinitely.  When the kernel  detects  such deadlocks, it causes one of the blocking lock requests to immediately fail with the error **EDEADLK**; an application that encounters such an error should release some of its locks to allow other applications to proceed before attempting regain the locks that  it requires.   Circular deadlocks involving more than two processes are also detected.  Note, however, that there are limitations to the kernel's deadlock-detection algorithm; see **BUGS**.

And in the BUGS section:

> ####  Deadlock detection
>
> The deadlock-detection algorithm employed by the kernel when dealing with **F_SETLKW** requests can yield both false negatives (failures to detect deadlocks, leaving a set of deadlocked processes blocked indefinitely) and false positives  (**EDEADLK**  errors  when     there is no deadlock).  For example, the kernel limits the lock depth of its dependency search to 10 steps, meaning that circular    deadlock chains that exceed that size will not be detected.  In addition, the kernel may falsely indicate a deadlock when two  or     more processes created using the **clone**(2) **CLONE_FILES** flag place locks that appear (to the kernel) to conflict.

I cannot say whether what you observe is a real deadlock, or a false positive from using **CLONE_FILES**.  Looking at `strace -f` output of the pytest run might be helpful to verify or eliminate that possibility.

**Comment by mgedmin** [comment]
> E   BlockingIOError: [Errno 35] Resource temporarily unavailable

This part baffles me somewhat, because "Resource temporarily unavailable" is the message for errno 11 (EAGAIN aka EWOULDBLOCK), not 35 (EDEADLK).

I see on the pytest bug that you've reproduced this both on Mac OS and on Debian.  Which OS is this for?  I'm only familiar with Linux.  If it comes from Mac OS, where errno numbers probably differ from Linux, please disregard my previous comment and read this one instead:

> #### F_SETLK (struct flock *)
> Acquire  a  lock  (when l_type is F_RDLCK or F_WRLCK) or release a lock (when l_type is F_UNLCK) on the bytes specified by the l_whence, l_start, and l_len fields of lock.  **If a conflicting lock is held by another process, this call  returns  -1 and  sets errno to EACCES or EAGAIN**.  (The error returned in this case differs across implementations, so POSIX requires a            portable application to check for both errors.)

(emphasis mine)

**Comment by mgedmin** [comment]
This issue is a great learning experience for me!

Turns out Python's `fcntl.flock()` [doesn't use fcntl(2)](https://docs.python.org/3/library/fcntl.html#fcntl.flock)!  It uses [flock(2)](https://manpages.debian.org/buster/manpages-dev/flock.2.en.html).  EWOULDBLOCK (aka EAGAIN) is the error for
> The file is locked and the LOCK_NB flag was selected.

LOCK_NB is what zc.lockfile explicitly uses to make the lock operation non-blocking.  I don't see that mentioned in zc.lockfile's documentation, which is IMHO a bug.

So: if the locking is non-blocking, then apparently you're supposed to write a retry loop yourself?  It would be nice if zc.lockfile provided a blocking locking primitive for conveniene, but that's a matter for a separate issue.

I think this one can be closed as "working as designed", but a new one (or two) issues should be opened instead:
- document that zc.lockfile.SimpleLockFile is non-blocking
- maybe request for a blocking option (maybe `with zc.lockfile.LockFile(..., timeout=60): ...`)

**Comment by StabbarN** [comment]
Thank you for your descriptive answers.

> Which OS is this for?

Both macOS and Linux.

> if the locking is non-blocking, then apparently you're supposed to write a retry loop yourself? 

That's how we solved it.
